rfcs/0003-mutations.md

```diff
@@ -0,0 +1,619 @@
+// the two different operation links (query & mutation)
+const link = split(
+  ({ query }) => {
+    const node = getMainDefinition(query);
+    return node.kind === "OperationDefinition" &&
+           node.operation === "mutation"
+  },
+  mutationLink,
+  queryLink
+)
+
+// Create an Apollo Client
+const client = new ApolloClient({
+  link,
+  cache: new InMemoryCache()
+})
+
+const CREATE_ENTITY = gql`
+  mutation createEntity($options: MyEntityOptions) {
+    createEntity(options: $options) {
+      id
+      name
+      value
+    }
+  }
+`
+
+// exec: execution function for the mutation query
+// loading: https://www.apollographql.com/docs/react/data/mutations/#tracking-mutation-status
+// state: mutation state instance
+const [exec, { loading, state }] = useMutation<MutationState<State>>(
+  CREATE_ENTITY,
+  {
+    client,
+    variables: {
+      options: { name: "...", value: 5 }
+    }
+  }
+)
+
+// Optimistic responses can be used to update
+// the UI before the execution has finished.
+// More information can be found here:
+// https://www.apollographql.com/docs/react/performance/optimistic-ui/
+const [exec, { loading, state }] = useMutation(
+  CREATE_ENTITY,
+  {
+    optimisticResponse: {
+      __typename: "Mutation",
+      createEntity: {
+        __typename: "MyEntity",
+        name: "...",
+        value: 5,
+        // NOTE: ID must be known so the
+        // final response can be correlated
+        id: "id"
+      }
+    },
+    variables: {
+      options: { name: "...", value: 5 }
+    }
+  }
+)
+```
+```html
+// Use the Mutation JSX Component
+<Mutation mutation={CREATE_ENTITY} variables={{options: { name: "...", value: 5 }}}>
+{(exec, { loading, state }) => (
+  <button onClick={exec} />
+)}
+</Mutation>
+```
+
+
+## Compatibility
+
+No breaking changes will be introduced, as mutations are an optional add-on to a subgraph.
+
+## Drawbacks and Risks
+
+Nothing apparent at the moment.
+
+## Alternatives
+
+The existing alternative that protocol developers are creating for dApp developers has been described above.
+
+## Open Questions
+
+- **How can mutations pickup where they left off in the event of an abrupt application shutdown?**
+  Since mutations can contain many different steps internally, it would be ideal to be able to support continuing resolver execution in the event the dApp abruptly shuts down.
+
+- **How can dApps understand what steps a given mutation will take during the course of its execution?**
+  dApps may way to present to the user friendly progress updates, letting them know a given action is 3/4 of the way through execution (for example) and a high level description of each step. I view this as closely tied to the previous open question above, as we could support continuing resolver executions if we know what step it's currently undergoing. More research will be done, and a potential solution to solve both of these use cases will be presented at a later date.
```

Comment:
It's separate from the concern of how a mutation can be "resumed".

I agree this is an improvement we can make in the future. I have an API in mind for this already though, where the state has a `steps: Step[]` field, where each step has the following shape:
```typescript
interface Step {
  id: string
  title: string
  description: string
  status: 'pending' | 'processing' | 'error' | 'finished',
  current: boolean
  error: Error?
  data: any
}
```
That, plus a few core events would be all we need to render UIs like those on https://ant.design/components/steps/.

But let's revisit this out of band and outside this RFC.

Reply:
Re "It's seperate...": Interesting, would love to hear your thoughts on how resuming a mutation could be implemented.

Re "I have an API in mind...": Awesome I really like this! Would love to implement this in the near future.

Should I remove this open question?

Reply:
I've added your suggested implementation to this open question.

Reply:
Great, thanks.

Re: resuming mutations. I think the best approach here would be to force resolvers into implementing a state machine that has a `restore` method or similar that takes a previously persisted state and identifies the state it should be in.

But for now I'd rather not force mutation resolvers into such a strict pattern. It could be a future add-on.